### ISSUE TITLE
Examples failed to compile with VSCode and GCC compiler

### ISSUE THREAD

**Issue by aregtech** [aregtech/areg-sdk#64]
**Describe the bug**
Try to compile codes in VSCode and GCC compiler of [20220701-candidate](https://github.com/aregtech/areg-sdk/tree/20220701-candidate) branch. All projects, except projects [10_locsvc](https://github.com/aregtech/areg-sdk/tree/20220701-candidate/examples/10_locsvc) and [11_locmesh](https://github.com/aregtech/areg-sdk/tree/20220701-candidate/examples/11_locmesh) are compiled and linked. The mentioned 2 projects failed during linking.

**To Reproduce**
Steps to reproduce the behavior.
- In VSCode select [CMakeLists.txt](https://github.com/aregtech/areg-sdk/blob/20220701-candidate/CMakeLists.txt) of the root folder and configure.
- After configuration, build all projects.

**Expected behavior**
All projects must be compiled and linked.

**Desktop or embedded (please complete the following information):**
- OS: Windows 10
- Hardware 64_x86
- Compiler cygwin-gcc
- AREG SDK version: [20220701-candidate](https://github.com/aregtech/areg-sdk/tree/20220701-candidate) branch.
 
**Additional context**
There were more projects that were not compiling, fixed by linking with libraries.
Also tried to compile projects with MS Visual Studio CMake tool, all projects compiled fine.

**Comment by Ali-Nasrolahi** [comment]
Thank you @aregtech for the report. :)
I'll take a look and keep this thread updated.

However, from the first glance; it's possible a few POSIX APIs that have used; have some incompatibility in Windows environment.

To be more specific I have encountered such issue by using `GCC 11.2.0 x86_64-w64-mingw32` option:
```c
[build] D:/areg-sdk/framework/areg/appbase/private/posix/ApplicationPosix.cpp: In function 'int {anonymous}::_getProcIdByName(const char*)':
[build] D:/areg-sdk/framework/areg/appbase/private/posix/ApplicationPosix.cpp:38:24: error: 'struct dirent' has no member named 'd_type'
[build]    38 |         if ((dirEntry->d_type == DT_REG) && (NEString::isNumeric<char>(dirEntry->d_name[0])))
[build]       |                        ^~~~~~
[build] D:/areg-sdk/framework/areg/appbase/private/posix/ApplicationPosix.cpp:38:34: error: 'DT_REG' was not declared in this scope
[build]    38 |         if ((dirEntry->d_type == DT_REG) && (NEString::isNumeric<char>(dirEntry->d_name[0])))
[build]       |                                  ^~~~~~
[build] make[3]: *** [CMakeFiles/areg.dir/build.make:137: CMakeFiles/areg.dir/framework/areg/appbase/private/posix/ApplicationPosix.cpp.obj] Error 1
[build] make[2]: *** [CMakeFiles/Makefile2:196: CMakeFiles/areg.dir/all] Error 2
[build] make[1]: *** [CMakeFiles/Makefile2:1109: CMakeFiles/01_hello.dir/rule] Error 2
[build] make: *** [Makefile:579: 01_hello] Error 2
```

Which as you can see is not even examples' issue anymore. The actual  `Areg`'s base having problems in compilation.

**Comment by aregtech** [comment]
@Ali-Nasrolahi , sorry. I forgot to attach the logs.

Here are the logs:
```log

[build] /usr/lib/gcc/x86_64-pc-cygwin/11/../../../../x86_64-pc-cygwin/bin/ld: CMakeFiles/11_locsvcmesh.dir/objects.a(ServiceClient.cpp.obj): in function `ProxyBase::hasNotificationListener(unsigned int) const':
[build] /cygdrive/e/Projects/aregtech/areg-sdk/build/E:/Projects/aregtech/areg-sdk/framework/areg/component/ProxyBase.hpp:1033: undefined reference to `ProxyBase::Listener::Listener(unsigned int, unsigned int)'
[build] /usr/lib/gcc/x86_64-pc-cygwin/11/../../../../x86_64-pc-cygwin/bin/ld: CMakeFiles/11_locsvcmesh.dir/objects.a(ServiceClient.cpp.obj): in function `bool __gnu_cxx::__ops::_Iter_equals_val<ProxyBase::Listener const>::operator()<__gnu_cxx::__normal_iterator<ProxyBase::Listener const*, std::vector<ProxyBase::Listener, std::allocator<ProxyBase::Listener> > > >(__gnu_cxx::__normal_iterator<ProxyBase::Listener const*, std::vector<ProxyBase::Listener, std::allocator<ProxyBase::Listener> > >)':
[build] /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/bits/predefined_ops.h:270: undefined reference to `ProxyBase::Listener::operator==(ProxyBase::Listener const&) const'
[build] /usr/lib/gcc/x86_64-pc-cygwin/11/../../../../x86_64-pc-cygwin/bin/ld: CMakeFiles/10_locservice.dir/objects.a(ServiceClient.cpp.obj): in function `ProxyBase::hasNotificationListener(unsigned int) const':
[build] /cygdrive/e/Projects/aregtech/areg-sdk/build/E:/Projects/aregtech/areg-sdk/framework/areg/component/ProxyBase.hpp:1033: undefined reference to `ProxyBase::Listener::Listener(unsigned int, unsigned int)'
[build] /usr/lib/gcc/x86_64-pc-cygwin/11/../../../../x86_64-pc-cygwin/bin/ld: CMakeFiles/10_locservice.dir/objects.a(ServiceClient.cpp.obj): in function `bool __gnu_cxx::__ops::_Iter_equals_val<ProxyBase::Listener const>::operator()<__gnu_cxx::__normal_iterator<ProxyBase::Listener const*, std::vector<ProxyBase::Listener, std::allocator<ProxyBase::Listener> > > >(__gnu_cxx::__normal_iterator<ProxyBase::Listener const*, std::vector<ProxyBase::Listener, std::allocator<ProxyBase::Listener> > >)':
[build] /usr/lib/gcc/x86_64-pc-cygwin/11/include/c++/bits/predefined_ops.h:270: undefined reference to `ProxyBase::Listener::operator==(ProxyBase::Listener const&) const'
[build] collect2: error: ld returned 1 exit status
[build] make[2]: *** [CMakeFiles/10_locservice.dir/build.make:137: E:/Projects/aregtech/areg-sdk/product/build/g++/Windows-x86_64-Debug/bin/10_locservice.exe] Error 1
[build] make[1]: *** [CMakeFiles/Makefile2:357: CMakeFiles/10_locservice.dir/all] Error 2
[build] make[1]: *** Waiting for unfinished jobs....
[build] collect2: error: ld returned 1 exit status
[build] make[2]: *** [CMakeFiles/11_locsvcmesh.dir/build.make:188: E:/Projects/aregtech/areg-sdk/product/build/g++/Windows-x86_64-Debug/bin/11_locsvcmesh.exe] Error 1
[build] make[1]: *** [CMakeFiles/Makefile2:410: CMakeFiles/11_locsvcmesh.dir/all] Error 2
[build] make: *** [Makefile:91: all] Error 2
[proc] The command: D:\Tools\CMake\bin\cmake.EXE --build e:/Projects/aregtech/areg-sdk/build --config Debug --target all -j 10 -- exited with code: 2 and signal: null
[build] Build finished with exit code 2
```
The logs of both projects refer to object [ProxyBase::LIstener](https://github.com/aregtech/areg-sdk/blob/ac17f06f25ae8aba57f09a0408fe3b55e24554dc/framework/areg/component/ProxyBase.hpp#L105). This is one of the most used objects. It exists in all other projects starting from example `10_locservice` and above. If that would cause a problem, then linking of all other projects should also fail. Probably some settings are wrong.

I addition, I've uploaded files that probably can be useful. Please download file [areg-sdk-issue-64.rar](https://drive.google.com/file/d/1R4IoTyXCjxio_AefHiSzxCR-fAY65yZk/view?usp=sharing). Inside there are 3 additional ZIP files:
1. `failing.zip` --> contains generated and compiled files of `10_locservice` and `11_locsvcmesh` projects. Maybe makes sense to see what is generated.
2. `succeed.zip` --> 2 additional project that succeed compilation. `12_pubsvc` because it has similar interface and `18_locwatchdog` because it uses so called `Local Service` as examples 10 and 11
3. `libs.zip` -- compiled libraries in case if you try manually to link.

Let me know if you need something else. We can as well organize session by video call and make tries on my machine. Feel free.

**Comment by Ali-Nasrolahi** [comment]
Thanks @aregtech :)

**Comment by aregtech** [comment]
@Ali-Nasrolahi , I found the reason why it is not compiling and made fixes:
1. The examples are not compiling because it requires to export `ProxyBase::Listener class` even if it is declared as `private`. The fix is `class AREG_API Listener`. After this change all projects are compiling. The compilation was broken only when compile 'areg' as shared library. I didn't analyze when exactly only examples 10 and 11, but it might have a reason. I guess, because there is an inline function call, which code is not included in library.
2. I analyzed and figured out why applications crash immediately after start. The problem is that the `gcc` compiler of `cygwin` does not resolve the dependencies of static objects. Some objects when instantiate, refer to another object, which is not instantiated yet. And this cause crash. The fix is if instead static members use static methods that return reference to object like this:
```cpp
const MyObject & MyClass::getMyObject()
{
    static const MyObject _myObject;
    return _myObject;
}
```
It is a known problem, which I forgot :)

Anyway, the problems are solved. Code is compiling, fixes are done. I make tests, will submit soon and close the ticket.
